File: companion/bolumler/bootstrap/ornek-01/cozum.py
import argparse
from collections import Counter
import csv
from itertools import combinations, product
import math
from pathlib import Path
import platform
import random
import statistics


BOOT_OLCU = ('n', 'ortalama', 'ampirik_varyans', 'orneklem_sd', 'kuramsal_boot_se',
             'yeniden_orneklem', 'boot_merkez', 'boot_varyans', 'boot_se',
             'boot_min', 'boot_max', 'alt025', 'ust975', 'boot_yanlilik', 'olasilik_toplami')
PERM_OLCU = ('n_a', 'n_b', 'ortalama_a', 'ortalama_b', 'gozlenen_fark', 'atama_sayisi',
             'uc_atama', 'p_tam', 'en_kucuk_fark', 'en_buyuk_fark', 'sifir_merkezi')


def csv_oku(yol, baslik, adet):
    with yol.open(encoding='utf-8-sig', newline='') as dosya:
        okuyucu = csv.reader(dosya)
        if next(okuyucu, []) != baslik:
            raise ValueError(f'Sutunlar uyusmuyor: {yol.name}')
        satirlar = list(okuyucu)
    if len(satirlar) != adet or any(len(satir) != len(baslik) for satir in satirlar):
        raise ValueError(f'Satir sayisi veya genisligi uyusmuyor: {yol.name}')
    return satirlar


def tamsayi(metin):
    deger = float(metin)
    if not math.isfinite(deger) or deger != int(deger):
        raise ValueError('Sonlu tam sayi gerekli.')
    return int(deger)


def gozlem_oku(yol, gruplu=False):
    baslik = ['id', 'grup', 'deger'] if gruplu else ['id', 'deger']
    adet = 6 if gruplu else 5
    satirlar = csv_oku(yol, baslik, adet)
    temiz = sorted((tamsayi(satir[0]), satir[1] if gruplu else '', float(satir[-1]))
                   for satir in satirlar)
    if [satir[0] for satir in temiz] != list(range(1, adet + 1)):
        raise ValueError('Kimlikler benzersiz ve 1..n olmali.')
    if any(not math.isfinite(satir[-1]) or abs(satir[-1]) > 1e12 for satir in temiz):
        raise ValueError('Bu ogretim paketi sonlu, mutlak degeri en cok 1e12 olan girdiler kabul eder.')
    if gruplu and Counter(satir[1] for satir in temiz) != Counter({'A': 3, 'B': 3}):
        raise ValueError('A ve B grubunda ucer gozlem gerekli.')
    return temiz


def planlar():
    bootstrap = list(product(range(1, 6), repeat=5))
    permutasyon = [tuple(secim) + tuple(kimlik for kimlik in range(1, 7) if kimlik not in secim)
                   for secim in combinations(range(1, 7), 3)]
    return bootstrap, permutasyon


def plan_oku(yol, beklenen):
    genislik = len(beklenen[0])
    satirlar = csv_oku(yol, ['sira'] + [f'ind{numara}' for numara in range(1, genislik + 1)], len(beklenen))
    temiz = sorted(tuple(tamsayi(alan) for alan in satir) for satir in satirlar)
    hedef = [(sira, *satir) for sira, satir in enumerate(beklenen, 1)]
    if temiz != hedef:
        raise ValueError(f'Tam sayim plani eksik veya degistirilmis: {yol.name}')
    return [satir[1:] for satir in temiz]


def girdiler(dizin):
    degerler = [satir[-1] for satir in gozlem_oku(dizin / 'veri.csv')]
    gruplar = gozlem_oku(dizin / 'permutasyon.csv', gruplu=True)
    bootstrap, permutasyon = planlar()
    return (degerler, gruplar,
            plan_oku(dizin / 'bootstrap-plan.csv', bootstrap),
            plan_oku(dizin / 'permutasyon-plan.csv', permutasyon))


def yuzdelik(degerler, olasilik):
    sirali = sorted(degerler)
    konum = (len(sirali) - 1) * olasilik
    alt = math.floor(konum)
    agirlik = konum - alt
    return sirali[alt] * (1 - agirlik) + sirali[min(alt + 1, len(sirali) - 1)] * agirlik


def hesapla(degerler, gruplar, bootstrap, permutasyon):
    ortalama = statistics.mean(degerler)
    ortalamalar = [statistics.mean(degerler[indis - 1] for indis in satir) for satir in bootstrap]
    merkez = statistics.mean(ortalamalar)
    ampirik_varyans = statistics.pvariance(degerler)
    boot_varyans = statistics.pvariance(ortalamalar)
    boot = (5, ortalama, ampirik_varyans, statistics.stdev(degerler),
            math.sqrt(ampirik_varyans / 5), len(ortalamalar), merkez, boot_varyans,
            math.sqrt(boot_varyans), min(ortalamalar), max(ortalamalar),
            yuzdelik(ortalamalar, .025), yuzdelik(ortalamalar, .975), merkez - ortalama,
            math.fsum(adet / len(ortalamalar) for adet in Counter(ortalamalar).values()))
    ortalama_a = statistics.mean(satir[-1] for satir in gruplar if satir[1] == 'A')
    ortalama_b = statistics.mean(satir[-1] for satir in gruplar if satir[1] == 'B')
    gozlenen = ortalama_a - ortalama_b
    tum = [satir[-1] for satir in gruplar]
    farklar = [statistics.mean(tum[indis - 1] for indis in satir[:3]) -
               statistics.mean(tum[indis - 1] for indis in satir[3:]) for satir in permutasyon]
    uc = sum(abs(fark) >= abs(gozlenen) - 1e-12 for fark in farklar)
    perm = (3, 3, ortalama_a, ortalama_b, gozlenen, len(farklar), uc, uc / len(farklar),
            min(farklar), max(farklar), statistics.mean(farklar))
    sonuc = [('bootstrap', olcu, deger) for olcu, deger in zip(BOOT_OLCU, boot)]
    sonuc += [('permutasyon', olcu, deger) for olcu, deger in zip(PERM_OLCU, perm)]
    return sonuc, ortalamalar, farklar


def kontrol_et(sonuc, yol):
    hedefler = csv_oku(yol, ['degisken', 'olcu', 'deger'], len(sonuc))
    for sonuc_satiri, hedef in zip(sonuc, hedefler):
        if list(sonuc_satiri[:2]) != hedef[:2]:
            raise ValueError('Referans etiketleri veya sirasi farkli.')
        deger = float(hedef[-1])
        if not math.isfinite(deger) or not math.isclose(sonuc_satiri[-1], deger, rel_tol=1e-9, abs_tol=1e-9):
            raise ValueError(f'Referans uyusmazligi: {hedef[1]}')
    print(f'DOGRULANDI: {len(sonuc)} kontrol degeri eslesiyor.')


def grafik_kaydet(dizin, ortalamalar, farklar, gozlenen):
    import matplotlib
    matplotlib.use('Agg')
    import matplotlib.pyplot as plt

    fig, eksenler = plt.subplots(1, 2, figsize=(11, 4))
    frekans = Counter(ortalamalar)
    eksenler[0].bar(sorted(frekans), [frekans[deger] / len(ortalamalar) for deger in sorted(frekans)],
                   width=.16, color='#1F4E79')
    for sinir in (yuzdelik(ortalamalar, .025), yuzdelik(ortalamalar, .975)):
        eksenler[0].axvline(sinir, color='#B45F06', linestyle='--')
    eksenler[0].set(title='3125 tam bootstrap orneklemi', xlabel='Ortalama', ylabel='Olasilik')
    frekans = Counter(farklar)
    renkler = ['#B45F06' if abs(deger) >= abs(gozlenen) - 1e-12 else '#1F4E79' for deger in sorted(frekans)]
    eksenler[1].bar(sorted(frekans), [frekans[deger] / len(farklar) for deger in sorted(frekans)],
                   width=.45, color=renkler)
    eksenler[1].set(title='20 tam etiket atamasi', xlabel='A - B ortalama farki', ylabel='Olasilik')
    fig.tight_layout()
    hedef = dizin / 'ciktilar/python'
    hedef.mkdir(parents=True, exist_ok=True)
    fig.savefig(hedef / 'bootstrap-permutasyon.png', dpi=150)
    plt.close(fig)


def main():
    parser = argparse.ArgumentParser(description='Tam bootstrap ve tam permutasyon: iki ayri ogretim ornegi')
    parser.add_argument('--check', action='store_true')
    parser.add_argument('--grafik', action='store_true')
    parser.add_argument('--benzetim', action='store_true')
    secenek = parser.parse_args()
    dizin = Path(__file__).parent
    veri = girdiler(dizin)
    sonuc, ortalamalar, farklar = hesapla(*veri)
    for degisken, olcu, deger in sonuc:
        print(f'{degisken:12} {olcu:22} {deger:.12g}')
    if secenek.check:
        kontrol_et(sonuc, dizin / 'beklenen-sonuclar.csv')
    if secenek.grafik:
        grafik_kaydet(dizin, ortalamalar, farklar, dict((olcu, deger) for _, olcu, deger in sonuc)['gozlenen_fark'])
    if secenek.benzetim:
        rng = random.Random(2026)
        tekrarlar = [statistics.mean(rng.choices(veri[0], k=5)) for tekrar in range(10000)]
        print('Monte Carlo (kontrol CSV disi):', statistics.stdev(tekrarlar),
              yuzdelik(tekrarlar, .025), yuzdelik(tekrarlar, .975))
    print('Python', platform.python_version())


if __name__ == '__main__':
    main()

File: companion/bolumler/bootstrap/ornek-01/cozum.R
gozlem_oku <- function(dosya, gruplu=FALSE) {
  veri <- read.csv(dosya, check.names=FALSE, stringsAsFactors=FALSE)
  baslik <- if (gruplu) c('id','grup','deger') else c('id','deger')
  adet <- if (gruplu) 6 else 5
  stopifnot(identical(names(veri),baslik), nrow(veri)==adet,
    is.numeric(veri$id), is.numeric(veri$deger), !anyNA(veri),
    all(is.finite(veri$id)), all(is.finite(veri$deger)), all(abs(veri$deger)<=1e12))
  veri <- veri[order(veri$id), ]
  stopifnot(all(veri$id==seq_len(adet)))
  if (gruplu) stopifnot(all(veri$grup %in% c('A','B')),
                        sum(veri$grup=='A')==3, sum(veri$grup=='B')==3)
  veri
}
plan_oku <- function(dosya, beklenen) {
  veri <- read.csv(dosya, check.names=FALSE)
  stopifnot(identical(names(veri),c('sira',paste0('ind',seq_len(ncol(beklenen))))),
            nrow(veri)==nrow(beklenen), all(vapply(veri,is.numeric,logical(1))),
            !anyNA(veri), all(is.finite(as.matrix(veri))))
  veri <- veri[order(veri$sira), ]
  stopifnot(all(veri$sira==seq_len(nrow(veri))), all(as.matrix(veri[,-1])==beklenen))
  as.matrix(veri[,-1])
}
veri <- gozlem_oku('veri.csv')
gruplar <- gozlem_oku('permutasyon.csv',TRUE)
beklenen_boot <- expand.grid(rep(list(1:5),5))
beklenen_boot <- as.matrix(beklenen_boot[do.call(order,beklenen_boot), ])
secimler <- t(combn(1:6,3))
beklenen_perm <- t(apply(secimler,1,function(indis) c(indis,setdiff(1:6,indis))))
boot_plan <- plan_oku('bootstrap-plan.csv',beklenen_boot)
perm_plan <- plan_oku('permutasyon-plan.csv',beklenen_perm)
ortalamalar <- apply(boot_plan,1,function(indis) mean(veri$deger[indis]))
ortalama <- mean(veri$deger)
merkez <- mean(ortalamalar)
ampirik_varyans <- mean((veri$deger-ortalama)^2)
boot_varyans <- mean((ortalamalar-merkez)^2)
boot <- c(5,ortalama,ampirik_varyans,sd(veri$deger),sqrt(ampirik_varyans/5),
          length(ortalamalar),merkez,boot_varyans,sqrt(boot_varyans),
          min(ortalamalar),max(ortalamalar),
          unname(quantile(ortalamalar,c(.025,.975),type=7)),
          merkez-ortalama,sum(table(ortalamalar)/length(ortalamalar)))
ortalama_a <- mean(gruplar$deger[gruplar$grup=='A'])
ortalama_b <- mean(gruplar$deger[gruplar$grup=='B'])
gozlenen <- ortalama_a-ortalama_b
farklar <- apply(perm_plan,1,function(indis)
  mean(gruplar$deger[indis[1:3]])-mean(gruplar$deger[indis[4:6]]))
uc <- sum(abs(farklar)>=abs(gozlenen)-1e-12)
perm <- c(3,3,ortalama_a,ortalama_b,gozlenen,length(farklar),uc,
          uc/length(farklar),min(farklar),max(farklar),mean(farklar))
sonuc <- data.frame(degisken=c(rep('bootstrap',15),rep('permutasyon',11)),
 olcu=c('n','ortalama','ampirik_varyans','orneklem_sd','kuramsal_boot_se',
        'yeniden_orneklem','boot_merkez','boot_varyans','boot_se','boot_min','boot_max',
        'alt025','ust975','boot_yanlilik','olasilik_toplami',
        'n_a','n_b','ortalama_a','ortalama_b','gozlenen_fark','atama_sayisi',
        'uc_atama','p_tam','en_kucuk_fark','en_buyuk_fark','sifir_merkezi'),deger=c(boot,perm))
print(sonuc,row.names=FALSE,digits=12)
secenek <- commandArgs(TRUE)
stopifnot(all(secenek %in% c('--check','--grafik','--benzetim')))
if ('--check' %in% secenek) {
  hedef <- read.csv('beklenen-sonuclar.csv',check.names=FALSE)
  stopifnot(identical(names(hedef),names(sonuc)),
    identical(hedef$degisken,sonuc$degisken),identical(hedef$olcu,sonuc$olcu),
    is.numeric(hedef$deger),all(is.finite(hedef$deger)),
    all(abs(hedef$deger-sonuc$deger)<=pmax(1e-9,abs(hedef$deger)*1e-9)))
  cat('DOGRULANDI: 26 kontrol degeri eslesiyor.\n')
}
if ('--grafik' %in% secenek) {
  dir.create('ciktilar/r',recursive=TRUE,showWarnings=FALSE)
  png('ciktilar/r/bootstrap-permutasyon.png',width=1600,height=600,res=150)
  par(mfrow=c(1,2))
  adet <- table(ortalamalar)
  plot(as.numeric(names(adet)),as.numeric(adet)/length(ortalamalar),type='h',
       xlab='Ortalama',ylab='Olasilik',main='3125 tam bootstrap orneklemi')
  abline(v=quantile(ortalamalar,c(.025,.975),type=7),lty=2,col='orange')
  adet <- table(farklar)
  plot(as.numeric(names(adet)),as.numeric(adet)/length(farklar),type='h',
       xlab='A - B farki',ylab='Olasilik',main='20 tam etiket atamasi')
  abline(v=c(-abs(gozlenen),abs(gozlenen)),lty=2,col='orange')
  dev.off()
}
if ('--benzetim' %in% secenek) {
  set.seed(2026)
  tekrarlar <- replicate(10000,mean(sample(veri$deger,5,replace=TRUE)))
  print(c(MC_se=sd(tekrarlar),quantile(tekrarlar,c(.025,.975),type=7)))
}
sessionInfo()